janusgraph-core/src/main/java/org/janusgraph/graphdb/olap/job/IndexRepairJob.java

```diff
@@ -182,18 +183,16 @@ public void process(JanusGraphVertex vertex, ScanMetrics metrics) {
                 assert indexType!=null;
                 IndexSerializer indexSerializer = graph.getIndexSerializer();
                 //Gather elements to index
-                List<JanusGraphElement> elements;
+                Iterable<JanusGraphElement> elements;
```

Comment:
I think it could be used as `Iterator` and traversed below because we access elements one by one only once below. 
```
Iterator<? extends JanusGraphElement> elements;
switch (indexType.getElement()) {
    case VERTEX:
        elements = Collections.singletonList(vertex).iterator();
        break;
    case PROPERTY:
        elements = addIndexSchemaConstraint(vertex.query(),indexType).properties().iterator();
        break;
    case EDGE:
        elements = addIndexSchemaConstraint(vertex.query().direction(Direction.OUT),indexType).edges().iterator();
        break;
    default: throw new AssertionError("Unexpected category: " + indexType.getElement());
}
```
And then below instead of traversing an Iterable we could traverse Iterator.
I.e. 
```
elements.forEachRemaining(element -> {
    if (indexSerializer.reindexElement(element, (MixedIndexType) indexType, documentsPerStore)) {
        metrics.incrementCustom(DOCUMENT_UPDATES_COUNT);
    }
});
```

Reply:
Good call. There really is no reason it can't be an Iterator, and that also avoids the type-safety issue that was causing the need to use `Iterables.transform(..., v -> v)` to unwrap and rewrap the element.